add network validation manager

simonvadee opened the pull request:
In this PR: 
- small improvement on the validators side, the list of validators is now indexed by the process name (no need to call `ShouldValidate` on all validators on the multiValidator).
- a NetworkManager implementing the `validation.Manager` interface was added.
- the subscription mechansim of the Manager has been factorized and splitted in its own file.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/stratumn/go-indigocore/415)
<!-- Reviewable:end -->

such commented:
I only made comments on code that was moved...
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCXzBfB6MVvLHV9uArb:bi26wz6) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
> ```Go
> 		b.listeners[index] = b.listeners[len(b.listeners)-1]
> 		b.listeners = b.listeners[:len(b.listeners)-1]
> 	}
> ```

it could make sense to store the `listeners` in a `map` with the `chan` as the key and an empty `struct` as the value (a `set` basically). It makes it easier to implement remove operations...

---

*[validation/broadcaster_test.go, line 30 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzX5M3q6TdHR0qZEa:-LCXzX5M3q6TdHR0qZEb:b99gprs) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L30)):*
> ```Go
> 
> 	t.Run("AddRemoveListener", func(t *testing.T) {
> 		t.Run("Adds a listener provided with the current valitor set", func(t *testing.T) {
> ```

nit: validator

---

*[validation/broadcaster_test.go, line 50 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzchX3VbEpbjXxc2Y:-LCXzchX3VbEpbjXxc2Z:b-xv4a4b) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L50)):*
> ```Go
> 		})
> 
> 		t.Run("Removes closes the channel", func(t *testing.T) {
> ```

I'm confused with the grammar here...

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415#-:-LCY0GzZ36fquQjHBhYq:b-eqfwom)*
<!-- Sent from Reviewable.io -->

alexppxela commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415#-:-LCY1C9wAXH1JGrQ66WM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

such commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCY6uVrBoyVko82bNgn:b-lpfoe6) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

But it makes it less efficient to iterate to broadcast, which is the primary operation of this object...I think this is a painful point of golang as a language (we need more built-in operations on slices/arrays) but we shouldn't ditch slices when they're the right tool for the job! Maybe it's time we compiled some slice helpers somewhere in our code to avoid re-typing this multiple times...
</blockquote></details>

Yeah you're right about iterating being the primary operation. I'd argue though that the right abstraction is the `set` (no duplication and we don't care about the order). And iterating over a hash map is not that expensive if the load factor is correctly adjusted.
Anyway we should definitely have those helpers!

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

t-bast commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCY8LvHDFsKaehIx78t:bfvmp9m) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

Yeah you're right about iterating being the primary operation. I'd argue though that the right abstraction is the `set` (no duplication and we don't care about the order). And iterating over a hash map is not that expensive if the load factor is correctly adjusted.
Anyway we should definitely have those helpers!
</blockquote></details>

I agree, the map's performance is probably enough for our usecase anyway :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

simonvadee commented:
Review status: 6 of 14 files reviewed at latest revision, 10 unresolved discussions.

---

*[validation/broadcaster.go, line 24 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzvQw4CdZzUCH_scH:-LCYF866Czj1YPI5XXAW:bvvy8uk) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L24)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

I suggest adding two interfaces to separate subscribers from the notification sender:

type UpdateTracker interface {
  func Subscribe() <-chan validators.Validator
  func Unsubscribe(<-chan validators.Validator)
}

type UpdateNotifier interface {
  func Broadcast(validator.Validator)
  func Close()
}

This concrete type would be named something like UpdateBroadcaster.
I think "Update" or something else is needed in the name otherwise it's really unclear what this notifies of.
And adding interfaces allows more mocking/testing if needed.
</blockquote></details>

Done. The validation.Manager interface is now composed with an `UpdateSubscriber`, but the `LocalManager` and `NetworkManager` wrap a concrete `UpdateBroadcaster` type, allowing them to access Broadcast and Close methods.

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCYc3gc8Qu1R88DfyAM:b98gpet) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

I agree, the map's performance is probably enough for our usecase anyway :)
</blockquote></details>

I went for the map but hit a wall : when storing a `map[chan something]struct{}`, the call to `delete(mymap, mychan)` cannot compile when `mychan` is of type `<-chan` (which is the case in `Unsubscribe(<-chan)`). I changed back to the slice, tell me if there's a better option !

---

*[validation/broadcaster_test.go, line 30 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzX5M3q6TdHR0qZEa:-LCY74q1EROuk2tUmvcB:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L30)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

nit: validator
</blockquote></details>

Done.

---

*[validation/broadcaster_test.go, line 39 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY1_EY792IZpf5h5-4:-LCYIx83BItgJv3FQDNT:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L39)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

nit: I think it's less confusing to use assert.* everywhere instead of mixing it with t.* calls
</blockquote></details>

Done.

---

*[validation/broadcaster_test.go, line 50 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzchX3VbEpbjXxc2Y:-LCYfAFfCxgS2-l_xtYU:bbaf64p) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L50)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

I'm confused with the grammar here...
</blockquote></details>

I added some comments, tell me if I missed the point.

---

*[validation/configloader.go, line 88 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY22dHAZEoMCPbmK3y:-LCY7Zq_BKOqHpkJGQuh:b-ky38rc) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/configloader.go#L88)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

nit: this shouldn't use errors.WithStack since it's an error returned by one of our own functions.
I think it's best to use errors.WithStack only to wrap errors coming from code we don't own.
</blockquote></details>

good point. Done.

---

*[validation/configloader.go, line 107 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY2Sv8-Qzbyu46KEFM:-LCY7vkA-DJzXbp0px6K:b-6zpirm) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/configloader.go#L107)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

nit: since we're dereferencing here, the function should directly receive a validators.PKI instead of a pointer (imo)
</blockquote></details>

I'm not sure  I agree, we need to check if the pointer is nil before dereferencing it, and this is the job of the `checkPKIConfig` function. 

---

*[validation/localmanager.go, line 46 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY3C-q-aHiNBI3pbB3:-LCYIzeY18cuIoV748ET:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/localmanager.go#L46)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

nit: it would be cleaner to use a constructor, especially if you introduce the interfaces I suggest. This part would only need to use the UpdateNotifier interface (not the UpdateSubscriber one).
</blockquote></details>

Done.

---

*[validation/networkmanager.go, line 47 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY3xvkEmpXBLiR0FKr:-LCY94cXALsl5gVZemyv:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/networkmanager.go#L47)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

nit: typo "NetworManager"
</blockquote></details>

Done.

---

*[validation/networkmanager.go, line 51 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY45oCCVav71zH5qra:-LCYJ-gC0961hsTQzNau:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/networkmanager.go#L51)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

This would only need an UpdateNotifier as well
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

such commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCYg94gA4jyfVJ7a08e:b2zihl7) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

I went for the map but hit a wall : when storing a `map[chan something]struct{}`, the call to `delete(mymap, mychan)` cannot compile when `mychan` is of type `<-chan` (which is the case in `Unsubscribe(<-chan)`). I changed back to the slice, tell me if there's a better option !
</blockquote></details>

I remember now that I actually hit the same wall a few weeks ago. I cannot remember if I found a workaround...

---

*[validation/broadcaster_test.go, line 50 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzchX3VbEpbjXxc2Y:-LCYgRdyET3bJK1k_Q30:b-5u6dmq) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L50)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

I added some comments, tell me if I missed the point.
</blockquote></details>

I actually meant that I don't understand why "close" and "remove" both end with an "s"... "Remove closes the channel" should be good expect that it's not really coherent with the other tests...

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

alexppxela commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[validation/localmanager.go, line 29 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYh5GX6aS0S2tWYRl0:-LCYh5GX6aS0S2tWYRl1:b-553vmh) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/localmanager.go#L29)):*
> ```Go
> // LocalManager manages governance for validation rules management in an indigo network.
> type LocalManager struct {
> 	*UpdateBroadcaster
> ```

You can put `UpdateNotifier` here?

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

simonvadee commented:
Review status: 13 of 14 files reviewed at latest revision, 5 unresolved discussions.

---

*[validation/broadcaster_test.go, line 50 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzchX3VbEpbjXxc2Y:-LCYjEuI-yLj4mt-off_:b-w4sbnw) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster_test.go#L50)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

I actually meant that I don't understand why "close" and "remove" both end with an "s"... "Remove closes the channel" should be good expect that it's not really coherent with the other tests...
</blockquote></details>

oh right ! better ?

---

*[validation/localmanager.go, line 29 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYh5GX6aS0S2tWYRl0:-LCYi9Tn4c_C4sYXVx3G:bqjz8ob) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/localmanager.go#L29)):*
<details><summary><i>Previously, alexppxela (Alexandre Thibault) wrote…</i></summary><blockquote>

You can put `UpdateNotifier` here?
</blockquote></details>

I can't, `LocalManager` needs to implement `Manager`, which implements `UpdateSubscriber` as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

t-bast commented:
Reviewed 8 of 8 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[validation/broadcaster.go, line 24 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzvQw4CdZzUCH_scH:-LCYhqtDAll8kfb9YUAx:b-ucfknh) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L24)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

Done. The validation.Manager interface is now composed with an `UpdateSubscriber`, but the `LocalManager` and `NetworkManager` wrap a concrete `UpdateBroadcaster` type, allowing them to access Broadcast and Close methods.
</blockquote></details>

Thanks!

---

*[validation/broadcaster.go, line 64 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCXzBfB6MVvLHV9uAra:-LCYi5pMEDhmH_7zub7E:b-1v730a) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/broadcaster.go#L64)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

I remember now that I actually hit the same wall a few weeks ago. I cannot remember if I found a workaround...
</blockquote></details>

Go isn't that simple in the end ;)

---

*[validation/broadcaster.go, line 20 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYhhgIC0n-w5niZoyP:-LCYhhgIC0n-w5niZoyQ:b-g1v86k) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/broadcaster.go#L20)):*
> ```Go
> 	"sync"
> 
> 	"github.com/stratumn/alice/core/protocol/coin/validator"
> ```

Not the validator you want I think :)

---

*[validation/configloader.go, line 107 at r1](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCY2Sv8-Qzbyu46KEFM:-LCYiR65-RKO0lbwl-Ov:b-u6jbsv) ([raw file](https://github.com/stratumn/go-indigocore/blob/8bf614082d9ebdb19f026195edf9fa8fe37fec36/validation/configloader.go#L107)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

I'm not sure  I agree, we need to check if the pointer is nil before dereferencing it, and this is the job of the `checkPKIConfig` function. 
</blockquote></details>

That makes sense, good point!

---

*[validation/localmanager.go, line 29 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYh5GX6aS0S2tWYRl0:-LCYiWW98y4UcksSnlQj:blybose) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/localmanager.go#L29)):*
<details><summary><i>Previously, alexppxela (Alexandre Thibault) wrote…</i></summary><blockquote>

You can put `UpdateNotifier` here?
</blockquote></details>

The trick is that the Manager interface only exposes the UpdateSubscriber part (which is good).
This component needs both (it uses an UpdateNotifier internally and exposes an UpdateSubscriber to clients).

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

alexppxela commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[validation/broadcaster.go, line 1 at r3](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYk9nt-p88Iszg9o6s:-LCYk9nt-p88Iszg9o6t:bnye4on) ([raw file](https://github.com/stratumn/go-indigocore/blob/dfff15f739e21e83feccd24e8f5bc7f5b47a4a10/validation/broadcaster.go#L1)):*
> ```Go
> // Copyright 2017 Stratumn SAS. All rights reserved.
> ```

We might update our readme and license file to 2018 (another PR)

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

simonvadee commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[validation/broadcaster.go, line 20 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYhhgIC0n-w5niZoyP:-LCYnYiq4uSj07132r4X:bx4ab5) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/broadcaster.go#L20)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

Not the validator you want I think :)
</blockquote></details>

I copy pasted the interfaces you wrote in the PR's comment, are you trying to sneak alice into indigocore and make the whole thing explode ? 
![shiet.gif](https://files.reviewableusercontent.io/images/reviewable/7048491/jX8U3pjOqxFYYBuu5iiIuEW9/shiet.gif)

---

*[validation/broadcaster.go, line 1 at r3](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYk9nt-p88Iszg9o6s:-LCYmnNd8MVtMlOt3RQl:bmx99xd) ([raw file](https://github.com/stratumn/go-indigocore/blob/dfff15f739e21e83feccd24e8f5bc7f5b47a4a10/validation/broadcaster.go#L1)):*
<details><summary><i>Previously, alexppxela (Alexandre Thibault) wrote…</i></summary><blockquote>

We might update our readme and license file to 2018 (another PR)
</blockquote></details>

cutting edge licenses !

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415)*
<!-- Sent from Reviewable.io -->

t-bast commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[validation/broadcaster.go, line 20 at r2](https://reviewable.io/reviews/stratumn/go-indigocore/415#-LCYhhgIC0n-w5niZoyP:-LCYoMbf2BKQ_AHdQv3h:b-g5urjb) ([raw file](https://github.com/stratumn/go-indigocore/blob/df8b96d99be84551e78a881d8b3fca1aac726647/validation/broadcaster.go#L20)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

I copy pasted the interfaces you wrote in the PR's comment, are you trying to sneak alice into indigocore and make the whole thing explode ? 
![shiet.gif](https://files.reviewableusercontent.io/images/reviewable/7048491/jX8U3pjOqxFYYBuu5iiIuEW9/shiet.gif)

</blockquote></details>

Bad, bad sneaky Alice...she'll have to be punished ;)

---


*Comments from [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/415#-:-LCYoUvsDeftFJu667Ym:bnfp4nl)*
<!-- Sent from Reviewable.io -->
